### ISSUE TITLE
Closure pointers are unstable

### ISSUE THREAD

**Issue by nfarina** [rehooks/local-storage#20]
Great library! I love the idea of keeping multiple localStorage hook consumers in sync, and I never knew about "custom events" so that's awesome.

I am seeing this error quite often:

```
Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application…
```

I believe this is happening because you are creating brand-new "wrapper" functions around `onLocalStorageChange` at multiple points. For instance, you sign up for events like this:

```js
window.addEventListener('storage', e => onLocalStorageChange(e));
```

And remove the listener later like this:

```js
window.removeEventListener('storage', e => onLocalStorageChange(e));
```

The trouble is, when you write `e => onLocalStorageChange(e)` you are creating a new function pointer that captures ("closes") over the current stack frame.

Toy example: `(e => {}) === (e => {})` gives `false`.

When `removeEventListener` is called, you're calling it with a different object, so it's never actually removed. It will keep getting called if the event fires, and when it's called it calls `updateLocalState` on an unmounted component which generates the error above.

The fix here is to use the React built-in `useCallback` hook to create a "stable" function pointer.

```js
const onLocalStorageChange = useCallback((event: Event) => { …logic… }, []);

// On mount:
window.addEventListener("storage", onLocalStorageChange);

// On unmount:
window.removeEventListener("storage", onLocalStorageChange);
```

Since `onLocalStorageChange` will always have the same value, it can be safely used to add/remove events. I can verify that this technique makes the above error go away.

**Comment by jharrilim** [comment]
Holy smokes! Neat find! We should definitely make this change. Would you like to make a PR? If not, we can probably get to it this weekend.

Thanks a lot!

**Comment by nfarina** [comment]
I would totally make a PR but I had to fork the library so I could make a version that did automatic JSON serialization. I posted my modified version so you could see what I did differently: https://gist.github.com/nfarina/807fdb7f63e4f010c198c446591db3e8

Hope it's helpful!

**Comment by jharrilim** [comment]
Updated: 10fef69

@iamsolankiamit Do you think we should automatically serialize objects to JSON? I have a feeling there will be more cases when users would rather have their object stored as opposed to `"[object Object]"`. We can probably just catch the circular JSON error and throw a better message so it is clear what happens when users try to pass in an object with circular references.

**Comment by iamsolankiamit** [comment]
Sure @jharrilim, it will reduce duplication in user land. Can you create a PR?

**Comment by lookfirst** [comment]
I see that someone added the useCallback, but they didn't update the way that the functions are passed in, so the problem persists. =(

window.removeEventListener('storage', e => onLocalStorageChange(e));
vs.
window.removeEventListener("storage", onLocalStorageChange);

**Comment by nfarina** [comment]
So speaking of unstable pointers…after an epic debugging session, I have come to discover that this line:

```js
if (event instanceof LocalStorageChanged) {
```

Will always be `false` on Safari. (It works fine in Chrome).

This check works on all browsers:

```js
if (event.type === LocalStorageChanged.eventName) {
```

**Comment by lookfirst** [comment]
Good find! (I don't use Safari.)

I don't even understand why that check exists... all that really matters is the second if statement... just add a check for the property existing... something like this (untested)...

```
			if (event.hasProperty('detail') && event.detail.key === key) {
				updateLocalState(event.detail.value);
			} else if (event.hasProperty('newValue') && event.key === key) {
				updateLocalState(event.newValue);
                         } else {
                                  throw Exception('unable to update local state');
                         }
```

**Comment by iamsolankiamit** [comment]
Fixed in [v2.2.2](https://github.com/rehooks/local-storage/releases/tag/v2.2.2)
